src/controller/scan/base_controller.go

```diff
@@ -233,78 +266,116 @@ func (bc *basicController) Scan(ctx context.Context, artifact *ar.Artifact, opti
+	if async {
+		go bc.startScanAll(bc.makeCtx(), executionID, async)
+	} else {
+		if err := bc.startScanAll(ctx, executionID, async); err != nil {
+			return 0, err
+		}
 	}
 
-	create := func(ctx context.Context, digest, registrationUUID, mimeType, trackID string, status job.Status) error {
-		reportPlaceholder := &scan.Report{
-			Digest:           digest,
-			RegistrationUUID: registrationUUID,
-			Status:           status.String(),
-			StatusCode:       status.Code(),
-			TrackID:          trackID,
-			MimeType:         mimeType,
-		}
-		// Set requester if it is specified
-		if len(ops.Requester) > 0 {
-			reportPlaceholder.Requester = ops.Requester
-		} else {
-			// Use the trackID as the requester
-			reportPlaceholder.Requester = trackID
-		}
+	return executionID, nil
+}
 
-		_, e := bc.manager.Create(reportPlaceholder)
-		return e
+func (bc *basicController) startScanAll(ctx context.Context, executionID int64, async bool) error {
+	if async {
+		// if async, this method is running in another goroutine ensure the execution exists in db
+		err := lib.RetryUntil(func() error {
+			_, err := bc.execMgr.Get(ctx, executionID)
+			return err
+		})
+		if err != nil {
+			log.Errorf("failed to get the execution %d for the scan all", executionID)
+			return err
+		}
 	}
 
-	if hasCapability(r, art) {
-		var producesMimes []string
-
-		for _, pm := range r.GetProducesMimeTypes(art.ManifestMediaType) {
-			if err = create(ctx, art.Digest, r.UUID, pm, trackID, job.PendingStatus); err != nil {
-				return "", nil, err
-			}
+	batchSize := 50
+	for artifact := range ar.Iterator(ctx, batchSize, nil, nil) {
+		scan := func(ctx context.Context) error {
+			return bc.Scan(ctx, artifact, WithExecutionID(executionID))
+		}
 
-			producesMimes = append(producesMimes, pm)
+		if err := orm.WithTransaction(scan)(ctx); err != nil {
+			// Just logged
+			log.Errorf("failed to scan artifact %s, error %v", artifact, err)
 		}
+	}
 
-		if len(producesMimes) > 0 {
-			return trackID, producesMimes, nil
+	count, err := bc.taskMgr.Count(ctx, q.New(q.KeyWords{"execution_id": executionID}))
+	if err != nil {
+		log.Errorf("failed to get the count of tasks for scan all, error %v", err)
+		return err
+	} else if count == 0 {
+		if err := bc.execMgr.MarkDone(ctx, executionID, "no task found"); err != nil {
+			log.Errorf("failed to mark the execution %d to be done, error: %v", executionID, err)
+			return err
 		}
 	}
 
-	err = create(ctx, art.Digest, r.UUID, v1.MimeTypeNativeReport, trackID, job.ErrorStatus)
-	return "", nil, err
+	return nil
 }
 
-func (bc *basicController) scanArtifact(ctx context.Context, r *scanner.Registration, artifact *ar.Artifact, trackID string, producesMimes []string) error {
-	jobID, err := bc.launchScanJob(ctx, trackID, artifact, r, producesMimes)
+func (bc *basicController) makeReportPlaceholder(ctx context.Context, r *scanner.Registration, art *ar.Artifact) ([]*scan.Report, error) {
```

Comment:
add more comments for this method? is it for report migration?

Reply:
It's not the report migration. For every artifact will be scanned, we will create a scan report for it before send the scan task to the jobservice.